src/pushsource/_impl/backend/koji_source.py

```diff
@@ -509,6 +519,104 @@ def _push_items_from_container_build(self, nvr, meta):
 
         return out
 
+    def _push_items_from_vmi_build(self, nvr, meta):
+        LOG.debug("Looking for virtual machine images on %s, %s", nvr, meta)
+
+        if not meta:
+            message = "Virtual machine image build not found in koji: %s" % nvr
+            LOG.debug(message)
+            raise ValueError(message)
+
+        if meta.get("container_koji_task_id"):
+            message = "The build %s is for container images." % nvr
+            LOG.debug(message)
+            raise ValueError(message)
+
+        # VMI has the {'typeinfo': {'image':{}}} on extra
+        extra = meta.get("extra") or {}
+        typeinfo = extra.get("typeinfo") or {}
+        image = typeinfo.get("image")
+
+        if image is None:
+            message = "Build %s not recognized as a virtual machine image build" % nvr
+            LOG.debug(message)
+            raise ValueError(message)
+
+        build_id = meta["id"]
+
+        archives = self._get_archives(build_id)
+
+        # Supported types of VMI archives
+        ami_types = ["raw", "raw-xz"]
+        azure_types = ["vhd", "vhd-compressed"]
+        cloud_types = {}
+
+        for k in ami_types:
+            cloud_types.setdefault(k, AmiPushItem)
+
+        for k in azure_types:
+            cloud_types.setdefault(k, VHDPushItem)
+
+        # Collect all image archives which are not containers
```

Comment:
It was raised earlier if it really makes sense to assume all of these archives are VM images.  Github marked that thread as Outdated but it still seems relevant.

Can we please not do this?  I don't know a lot about koji, but I get the sense there are not very strict rules around what is recorded as an "image".  I see no reason it would be valid to assume that all images are VM images by default.  There are already counterexamples:

- the container case (which is already handled in the code)
- CD/DVD ISO images used for installers, e.g. https://url.corp.redhat.com/76d5209

Though you *can* try to write code of the form "assume they're all VM images except those we explicitly filter out", as currently done for containers: even if such code would work today, on the koji side there is apparently no requirement that 'image' archives are VM images.  So new types of image archives could be added at any time, making the code suddenly become wrong.

I think it should be limited to only those you have some more specific reason to believe are VM images.  If we can only do that by matching file extensions or similar, that's unfortunate but still better than assuming everything is a VM image.

Note, it's always easier to add support for more types later (which is backwards-compatible) than it is to remove something which shouldn't have been supported (which is potentially not backwards-compatible).  It's wise to start with supporting a narrow set of inputs and broaden it as we learn more.

Reply:
Thanks for bringing this out again, IDK why it was marked as `outdated` as well.

I've updated the code to only match the known extensions from `type_name` through `cloud_types.keys()`. With this we should only include the expected VM images.

PTAL again